### DISCUSSION TITLE
How do I set the current version of PHP in use at upgrade time?

### DISCUSSION THREAD

**Discussion by Yi-pixel** [rectorphp/rector#7863] [Q&A] [answered]
Hi!

The current project code is running on PHP 7.1 and I plan to upgrade to PHP 8.1

I installed rector in PHP 7.1 CLI, composer installed `0.12.*` version (old version) for me, then I added to rector.php。

- rector.php

```php
<?php


declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Rector\Set\ValueObject\LevelSetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [
        __DIR__ . '/src',
    ]);

    $parameters->set(Option::PHP_VERSION_FEATURES, \Rector\Core\ValueObject\PhpVersion::PHP_81);

    // Define what rule sets will be applied
    $containerConfigurator->import(LevelSetList::UP_TO_PHP_81, null, true);
};

```

However, the process is too long due to the very large amount of code.

And it uses too much memory, about 20G or so.

Then I noticed that rector `0.15.*` (newer) can be installed with PHP 7.2.

So I'm going to upgrade my code to PHP 7.2 first, then install the new rector, and then upgrade to PHP 7.4 or PHP 8.1.

---

However, there is a problem if I upgrade to PHP 7.2 and then use the `\Rector\Set\ValueObject\LevelSetList::UP_TO_PHP_81` rule set.

I find that these rule sets are recursive, i.e. it automatically contains PHP 7.4 rules, which in turn contain 7.3.

Even more.

---

I noticed that the upgrade_to_xx rule set

7.3 includes 7.2
7.2 contains 7.1
7.1 contains 7.0 
7.0 contains 5.6

So if the current code is already at PHP 7.1, then if I set upgrade_to_73, will the rector apply the rules below 7.1, such as PHP71, PHP70, PHP56...?

---

Now I have some questions:

- Can I set the current(source) PHP version of my code? I noticed there is a `PHP_VERSION_FEATURES` is this? That is, if I'm currently on PHP 7.1, I set `PHP_VERSION_FEATURES` to `\Rector\Core\ValueObject\PhpVersion::PHP_71`.

This way rector may no longer automatically apply the rules of PHP 7.1 and below

- Or can I only do it the `SetList::PHP_72`, `SetList::PHP_73`, `SetList::PHP_74` way for each version?

- Also, rector version 0.12 doesn't seem to include all 8.1 features, so I still have to upgrade to 0.12 (PHP 72) first, right?

---

Thank you ☺

**Comment by samsonasik** [comment, marked as answer]
You can apply specific set list, one by one for that, see https://getrector.com/documentation/set-lists

**Reply by Yi-pixel** [reply]
Thanks
